Create a sticky TDI session to reuse for multiple writes to pipeline

5abeel opened the pull request:
A performance improvement suggested and tested by IDC team

5abeel committed:
Create a sticky TDI session to reuse for multiple writes to pipeline

Signed-off-by: Sabeel Ansari <sabeel.ansari@intel.com>

ffoulkes reviewed (changes requested):
I see several problems with this design:

1. **It uses static variables**. Except in certain specific circumstances, static variables are inconsistent with the design of Stratum. They introduce global state, and they make the code non-reentrant and more difficult to test.
   - The solution to this is straightforward: make the saved pointer to the persistent session an instance variable of the es2k_node class.
   -PR (https://github.com/ipdk-io/stratum-dev/pull/228) makes this change to the current PR. (merged)
2. **It is an unconditional change**. A persistent forwarding session may be appropriate in this case, but is it appropriate in all cases? And if it proves wrong, there's no way to disable the feature without editing and rebuilding the source code.
   - My inclination would be to introduce a "persistent forwarding session" Boolean flag to enable this feature. Its default value should be `false`, so the code is backward compatible by default.
   - PR https://github.com/ipdk-io/stratum-dev/pull/229 makes this change to the current PR. (merged)
3. **It makes no provision for concurrency.** What happens if more than one client sends requests to the node? At the same time?
   - Stratum uses an Abseil Mutex (the `lock_` instance variable) to protect access to node-specific state. I haven't looked into it in depth, but it might be applicable for this variable, too.
   - **Not a problem:** It turns out that `WriteForwardingEntries` already takes a lock on the node object on entry.

ffoulkes committed:
Replace static variables with instance variable (#228)

Signed-off-by: Derek Foster <derek.foster@intel.com>

ffoulkes committed:
Implement flag to enable sticky TDI sessions (#229)

- Implemented an --enable_sticky_tdi_session flag to enable the
  use of a persistent TDI session for WriteForwardingEntries
  requests. It is disabled (false) by default.

Signed-off-by: Derek Foster <derek.foster@intel.com>

ffoulkes reviewed (approved):
LGTM once the revised code has been tested.

I'd feel better if someone else signed off on this as well, as I am now effectively the author of the code. 😲

ffoulkes commented:
For the record:

- I successfully built the code for ES2K using Bazel and ran the unit test suite.
- The CI pipeline successfully built the code for DPDK using Bazel and ran the unit test suite.

The fact that the unit test suite ran successfully is not a strong indication of correctness. We do not have a unit test for the Node class.
